presto-main/src/main/java/com/facebook/presto/memory/DefaultQueryContext.java

```diff
@@ -196,7 +196,11 @@ private synchronized boolean tryUpdateUserMemory(String allocationTag, long delt
     {
         if (delta <= 0) {
             ListenableFuture<?> future = updateUserMemory(allocationTag, delta);
-            verify(future.isDone(), "future should be done");
+            // When delta == 0 and the pool is full the future can still not be done,
```

Comment:
why call `updateUserMemory` at all when delta==0?
now, that we (implicitly) special-case delta==0, it makes sense to handle it more properly

Reply:
Yes, we need to handle the delta = 0 case more properly, for that I will create a subsequent PR( with that I will also address your comment below).

Reply:
> why call updateUserMemory at all when delta==0?

Looking at the code again, looks like we still need to call `updateUserMemory` otherwise listeners will not be called in the memory pools (see `MemoryPool:onMemoryReserved()`), which will change the behavior of the current implementation.

Reply:
I didn't look into the code, so i might be missing something... Why do we want to call listeners if allocations didn't change?

BTW, if we special case `delta==0`, we can probably do this on lower-level contexts too, to avoid calling synchronized methods.

Reply:
> I didn't look into the code, so i might be missing something... Why do we want to call listeners if allocations didn't change?

`MemoryPool.reserve()` has always been called with `0` bytes, and looking at the git history the listeners are added to the `MemoryPool` is with e83c05cf7b0c4c70ed0e423d77d4f08f510efe1c, so we started calling those listeners for `delta = 0` starting with that change. Currently, I see that only `MemoryRevokingScheduler` relies on these listeners. Since you know that class better than me, if you think it's safe to not call listeners for 0 bytes, then I think we can special case `delta==0` at the lower-level contexts yes.

Reply:
@nezihyigitbasi thanks for looking in this.
Yes, from `MemoryRevokingScheduler` / e83c05cf7b0c4c70ed0e423d77d4f08f510efe1c perspective, no change in memory allocation (`delta==0`), is not an interesting event.
And, I think it is a generally reasonable and future-safe assumption that `MemoryPoolListener#onMemoryReserved` does not need to  (even: should not) be called when no new memory is reserved.